pkg/scaling/scale_handler_raw_metrics.go

```diff
@@ -0,0 +1,186 @@
+	rawMetricsChannelCapacity = 16
+
+	// timeout after which all the metrics for the subscriber will be unsubscribed (if these are not being consumed by client)
+	unsubscribeTimeout = 1 * time.Minute
+)
+
+func (h *scaleHandler) UnsubscribeMetric(_ context.Context, subscriber string, sor *api.ScaledObjectRef) bool {
+	toDel := toMetricMeta(sor)
+	log.Info("Unsubscribing metric", "subscriber", subscriber, "metric", toDel)
+	h.subsLock.Lock()
+	defer h.subsLock.Unlock()
+	if _, found := h.rawMetricsSubscriptions[subscriber]; !found || !h.isSubscribed(subscriber, sor) {
+		log.V(10).Info("Unsubscribing metric - noop")
+		return false
+	}
+
+	updatedSlice := slices.DeleteFunc(h.rawMetricsSubscriptions[subscriber].subscriptions, func(item metricMeta) bool {
+		return item.equal(toDel)
+	})
+	h.rawMetricsSubscriptions[subscriber].subscriptions = updatedSlice
+	if len(updatedSlice) == 0 {
+		close(h.rawMetricsSubscriptions[subscriber].rawMetrics)
+		select {
+		case h.rawMetricsSubscriptions[subscriber].done <- true:
+		default:
+		}
+		close(h.rawMetricsSubscriptions[subscriber].done)
+	}
+	mm := h.rawMetricsSubscriptions[subscriber].subscriptions
+	log.V(10).Info("All subscribed metrics", "subscriber", subscriber, "subscribersMetrics", mm)
+	delete(h.rawMetricsSubscriptions, subscriber)
+	return true
+}
+
+func (h *scaleHandler) SubscribeMetric(_ context.Context, subscriber string, sor *api.ScaledObjectRef) bool {
+	toAdd := toMetricMeta(sor)
+	log.Info("Subscribing metric", "subscriber", subscriber, "metric", toAdd)
+	h.subsLock.Lock()
+	defer h.subsLock.Unlock()
+	if _, found := h.rawMetricsSubscriptions[subscriber]; !found {
+		h.rawMetricsSubscriptions[subscriber] = makeEmptySubscriptions(subscriber)
+	} else if h.isSubscribed(subscriber, sor) {
+		log.V(10).Info("Subscribing metric - noop")
+		return true
+	}
+
+	h.rawMetricsSubscriptions[subscriber].subscriptions = append(h.rawMetricsSubscriptions[subscriber].subscriptions, toAdd)
+	mm := h.rawMetricsSubscriptions[subscriber].subscriptions
+	log.V(10).Info("All subscribed metrics", "subscriber", subscriber, "subscribersMetrics", mm)
+	return false
+}
+
+func (h *scaleHandler) GetRawMetricsChan(subscriber string) (chan RawMetrics, chan bool) {
+	h.subsLock.Lock()
+	defer h.subsLock.Unlock()
+	if _, found := h.rawMetricsSubscriptions[subscriber]; !found {
+		h.rawMetricsSubscriptions[subscriber] = makeEmptySubscriptions(subscriber)
+	}
+	return h.rawMetricsSubscriptions[subscriber].rawMetrics, h.rawMetricsSubscriptions[subscriber].done
+}
+
+func (h *scaleHandler) sendWhenSubscribed(soName, ns, triggerName string, metrics []external_metrics.ExternalMetricValue) {
+	mm := metricMeta{
+		TriggerName:      triggerName,
+		ScaledObjectName: soName,
+		Namespace:        ns,
+	}
+	h.subsLock.Lock()
+	targets := make([]*RawMetricSubscriptions, 0, len(h.rawMetricsSubscriptions))
+	for _, s := range h.rawMetricsSubscriptions {
```

Comment:
perf nit: the way it's currently designed, it's efficient to add or remove a subscriber but costly to check if the metric should be forwarded or not, because for each metric, it needs to iterate over each subscriber and each combination of SO+trigger defined in their list. For small environments it doesn't matter but it may add up for 100+ subscriptions.

Because I suspect the number of subscriber + hso + trigger will be much smaller than the overall number of metrics, how about adding a quick lookup map just to check if the combination is not subscribed and can be dropped? e.g.
```
map["so.Namespace/so.Name/triggerName"][]subscriber_name
```
with a single lookup, it can see if the metric doesn't need forwarding and also when it does need it, it can see directly all subscribers for it

Reply:
added a metric (ns+so+triggerName) to subscriptions map so that sendWhenSubscribed will be fast. This made the (un)subscribe methods little bit more complicated, but these are not going to be called that often as `sendWhenSubscribed`